x/evm/types/statedb.go

```diff
@@ -609,6 +610,25 @@ func (csdb *CommitStateDB) Snapshot() int {
 	return id
 }
 
+//GetRevisionID return the revision id from store, return value will be zero if revision id not exists in store
+func (csdb *CommitStateDB) GetRevisionID() int {
+	store := prefix.NewStore(csdb.ctx.KVStore(csdb.storeKey), KeyPrefixRevisionID)
+	bz := store.Get(RevisionKey()) //if not exist, that means RevisionId was never used, return 0
+	if len(bz) == 0 {
+		return 0
+	}
+
+	revision := binary.BigEndian.Uint64(bz)
+	return int(revision)
+}
+
+//set newest revision id to store
+func (csdb *CommitStateDB) SetRevisionID() {
+	revisionID := csdb.GetRevisionID() + csdb.nextRevisionID
```

Comment:
I don't understand this addition

Reply:
let me explain it:
* now `csdb.nextRevisionID` is a memory cache
* we store `Revision ID` in every call of EndBlock func
* GetRevisionID() will only return the last available ID
* `csdb.nextRevisionID` will increased in function`Snapshot ` but value based `csdb.GetRevisionID()` such as:
```go
func (csdb *CommitStateDB) Snapshot() int {
	id := csdb.nextRevisionID
	id := csdb.nextRevisionID + csdb.GetRevisionID()
	csdb.nextRevisionID++
```
* when we store the revisionID, we are going to store the last available ID, it must include current ` csdb.nextRevisionID`
* so here need add the csdb.nextRevisionID

Reply:
Another thing is we can not store revision id in every time the `csdb.nextRevisionID` changed in function `Snapshot`,  for example we assume that:
* block height = 10, include 20 numbers of EVM transactions , revision id = 5
* executing block h:10
* when we executing the tx of index 11 in block h:10, the revision id will be 16
* if we save 16 to store
* assume that current node has stopped and restart
* block h:10 will be re-executed, but now, the local stored revision id is 16, in-fact what we need the revision id is 5
* so wrong revision id received
that's why I saved revision id in `EndBlock`